v21.2/build-a-nodejs-app-with-cockroachdb-prisma.md

```diff
@@ -0,0 +1,167 @@
+    <a href="build-a-nodejs-app-with-cockroachdb.html"><button class="filter-button">Use <strong>node-postgres</strong></button></a>
+    <a href="build-a-nodejs-app-with-cockroachdb-sequelize.html"><button class="filter-button">Use <strong>Sequelize</strong></button></a>
+    <a href="build-a-nodejs-app-with-cockroachdb-knexjs.html"><button class="filter-button">Use <strong>KnexJS</strong></button></a>
+    <a href="build-a-nodejs-app-with-cockroachdb-prisma.html"><button class="filter-button current">Use <strong>Prisma</strong></button></a>
+    <a href="build-a-typescript-app-with-cockroachdb.html"><button class="filter-button">Use <strong>TypeORM</strong></button></a>
+</div>
+
+This tutorial shows you how build a simple Node.js application with CockroachDB and [Prisma Client](https://www.prisma.io).
+
+## Step 1. Start CockroachDB
+
+{% include {{ page.version.version }}/app/sample-setup.md %}
+
+## Step 2. Get the code
+
+Clone the code's GitHub repo:
+
+{% include_cached copy-clipboard.html %}
+~~~ shell
+$ git clone https://github.com/cockroachlabs/example-app-node-prisma
+~~~
+
+The project has the following directory structure:
+
+~~~
+├── dbinit.sql
+├── index.js
+├── package.json
+└── prisma
+    └── schema.prisma
+~~~
+
+The `dbinit.sql` file initializes the database schema that the application uses:
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+{% remote_include https://raw.githubusercontent.com/cockroachlabs/example-app-node-prisma/main/dbinit.sql %}
+~~~
+
+The `index.js` file contains the code for `INSERT`, `SELECT`, `UPDATE`, and `DELETE` SQL operations:
+
+{% include_cached copy-clipboard.html %}
+~~~ js
+{% remote_include https://raw.githubusercontent.com/cockroachlabs/example-app-node-prisma/main/index.js %}
+~~~
+
+All of the database operations are wrapped in a helper function named `retryTxn`. This function attempts to commit statements in the context of an explicit transaction. If a [retry error](transaction-retry-error-reference.html) is thrown, the wrapper will retry committing the transaction, with [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff), until the maximum number of retries is reached (by default, 15).
+
+## Step 3. Initialize the database
+
+{{site.data.alerts.callout_info}}
+CockroachDB does not support Prisma Migrate. We recommend executing DDL SQL statements to initialize your database schema separately. After the database schema is initialized, you can load existing tables into your Prisma schema with the Prisma Client, as demonstrated [here](https://www.prisma.io/docs/concepts/components/introspection).
+{{site.data.alerts.end}}
+
+To initialize the example database schema, use the [`cockroach sql`](cockroach-sql.html) command to execute the SQL statements in the `dbinit.sql` file:
+
+<div class="filter-content" markdown="1" data-scope="cockroachcloud">
+
+{% include_cached copy-clipboard.html %}
+~~~ shell
+cat dbinit.sql | cockroach sql --url "<connection-string>"
```

Comment:
This might not work on Windows? It should be more cross-platform to use the `--file` flag with `cockroach sql`.

Reply:
I did test it in Windows, yes.

IIRC you needed a v21.1 or later binary to use `--file` and for whatever reason many Windows users had an earlier, incompatible binary. So to make a single command that worked across all OSes we used `cat`.

If we think most users have an updated v21.1+ binary, or can use this as an opportunity to get them to upgrade, I do think using `--file` is the more correct option.

Reply:
My reasoning was based on my assumption that Windows does not come with the `cat` command, so it would require an extra installation step.